File: src/main/python/g0301_0400/s0394_decode_string/Solution.py
# #Medium #Top_100_Liked_Questions #String #Stack #Recursion #LeetCode_75_Stack
# #Level_1_Day_14_Stack #Udemy_Strings #Big_O_Time_O(n)_Space_O(n)
# #2025_07_25_Time_0_ms_(100.00%)_Space_17.56_MB_(97.37%)

class Solution:
    def __init__(self):
        self.i = 0

    def decodeString(self, s: str) -> str:
        count = 0
        sb = []
        while self.i < len(s):
            c = s[self.i]
            self.i += 1
            if c.isalpha():
                sb.append(c)
            elif c.isdigit():
                count = count * 10 + int(c)
            elif c == ']':
                break
            elif c == '[':
                # sub problem
                repeat = self.decodeString(s)
                sb.append(repeat * count)
                count = 0
        return ''.join(sb)

File: src/main/php/g0301_0400/s0394_decode_string/Solution.php
<?php

namespace leetcode\g0301_0400\s0394_decode_string;

// #Medium #Top_100_Liked_Questions #String #Stack #Recursion #LeetCode_75_Stack
// #Level_1_Day_14_Stack #Udemy_Strings #Big_O_Time_O(n)_Space_O(n)
// #2023_12_24_Time_3_ms_(77.78%)_Space_19.7_MB_(11.11%)

class Solution {
    private $i = 0;

    /**
     * @param String $s
     * @return String
     */
    public function decodeString($s) {
        $count = 0;
        $sb = "";
        while ($this->i < strlen($s)) {
            $c = $s[$this->i];
            $this->i++;
            if (ctype_alpha($c)) {
                $sb .= $c;
            } elseif (ctype_digit($c)) {
                $count = $count * 10 + intval($c);
            } elseif ($c == ']') {
                break;
            } elseif ($c == '[') {
                // sub problem
                $repeat = $this->decodeString($s);
                while ($count > 0) {
                    $sb .= $repeat;
                    $count--;
                }
            }
        }
        return $sb;
    }
}
